File: ios/Sources/FacebookAppEventsPlugin/FacebookAppEventsPlugin.swift
import Foundation
import Capacitor
import FBSDKCoreKit
import AppTrackingTransparency
import AdSupport

@objc(FacebookAppEventsPlugin)
public class FacebookAppEventsPlugin: CAPPlugin , CAPBridgedPlugin {
    public let identifier = "FacebookAppEventsPlugin"
    public let jsName = "FacebookAppEvents"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "initialize", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isInitialized", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "logEvent", returnType: CAPPluginReturnPromise)

    ]

    private var initialized = false
    private static let TAG = "FacebookAppEvents"

 @objc func initialize(_ call: CAPPluginCall) {
        guard let appId = call.getString("appId"),
              let clientToken = call.getString("clientToken") else {
            call.reject("Missing or empty appId or clientToken")
            return
        }

        if appId.isEmpty {
            call.reject("Missing or empty appId")
            return
        }

        if clientToken.isEmpty {
            call.reject("Missing or empty clientToken")
            return
        }

        do {
            // Configure Facebook SDK
            Settings.shared.appID = appId
            Settings.shared.clientToken = clientToken
            Settings.shared.isAutoInitEnabled = true
            Settings.shared.isAutoLogAppEventsEnabled = true

            // Enable debug logging in debug builds
            #if DEBUG
            Settings.shared.isLoggingBehaviorEnabled = true
            Settings.shared.loggingBehaviors = [.appEvents]
            #endif

            // Request tracking authorization for iOS 14+
            if #available(iOS 14, *) {
                ATTrackingManager.requestTrackingAuthorization { status in
                    switch status {
                    case .authorized:
                        let idfa = ASIdentifierManager.shared().advertisingIdentifier
                        print("\(Self.TAG): IDFA authorized: \(idfa)")
                    case .denied, .restricted, .notDetermined:
                        print("\(Self.TAG): Tracking authorization not granted")
                    @unknown default:
                        break
                    }
                }
            }

            // Initialize SDK
            ApplicationDelegate.shared.initializeSDK()
            
            // Activate app
            AppEvents.shared.activateApp()

            initialized = true
            print("\(Self.TAG): Facebook SDK initialized successfully")
            call.resolve()
        } catch {
            print("\(Self.TAG): Failed to initialize Facebook SDK: \(error)")
            call.reject("Failed to initialize Facebook SDK: \(error.localizedDescription)")
        }
    }

    @objc func isInitialized(_ call: CAPPluginCall) {
        let isSDKInitialized = Settings.shared.appID != nil && !Settings.shared.appID!.isEmpty
        call.resolve(["initialized": initialized && isSDKInitialized])
    }

    @objc func logEvent(_ call: CAPPluginCall) {
        if !initialized {
            call.reject("Facebook SDK not initialized. Call initialize() first.")
            return
        }

        if !ensureSDKInitialized() {
            call.reject("Facebook SDK not properly initialized")
            return
        }

        guard let eventName = call.getString("eventName") else {
            call.reject("Missing or empty 'eventName'")
            return
        }

        if eventName.isEmpty {
            call.reject("Missing or empty 'eventName'")
            return
        }

        let valueToSum = call.getDouble("valueToSum")
        let parameters = call.getObject("parameters") ?? [:]

        do {
            let fbParams = try createParametersFromDictionary(parameters)

            // Handle standard Facebook events with special processing
            if isPurchaseEvent(eventName) {
                try logPurchaseEvent(eventName: eventName, valueToSum: valueToSum, parameters: fbParams, call: call)
            } else {
                logStandardEvent(eventName: eventName, valueToSum: valueToSum, parameters: fbParams)
                call.resolve()
            }
        } catch {
            print("\(Self.TAG): Error logging event: \(eventName), error: \(error)")
            call.reject("Failed to log event: \(error.localizedDescription)")
        }
    }

    private func createParametersFromDictionary(_ parameters: [String: Any]) throws -> [AppEvents.ParameterName: Any] {
        var fbParams: [AppEvents.ParameterName: Any] = [:]

        for (key, value) in parameters {
            let parameterName = AppEvents.ParameterName(rawValue: key)
            
            if let stringValue = value as? String {
                fbParams[parameterName] = stringValue
            } else if let intValue = value as? Int {
                fbParams[parameterName] = intValue
            } else if let doubleValue = value as? Double {
                fbParams[parameterName] = doubleValue
            } else if let boolValue = value as? Bool {
                fbParams[parameterName] = boolValue
            } else if let floatValue = value as? Float {
                fbParams[parameterName] = floatValue
            } else if let longValue = value as? Int64 {
                fbParams[parameterName] = longValue
            } else {
                // Convert other types to string as fallback
                fbParams[parameterName] = String(describing: value)
            }
        }

        return fbParams
    }

    private func isPurchaseEvent(_ eventName: String) -> Bool {
        return eventName == "fb_mobile_purchase" || 
               eventName == "Purchase" ||
               eventName == AppEvents.Name.purchased.rawValue
    }

    private func logPurchaseEvent(eventName: String, valueToSum: Double?, parameters: [AppEvents.ParameterName: Any], call: CAPPluginCall) throws {
        // For purchase events, we need currency information
        var currencyCode: String?
        
        if let fbCurrency = parameters[AppEvents.ParameterName(rawValue: "fb_currency")] as? String {
            currencyCode = fbCurrency
        } else if let currency = parameters[AppEvents.ParameterName(rawValue: "currency")] as? String {
            currencyCode = currency
        }

        guard let currency = currencyCode, !currency.isEmpty else {
            call.reject("'fb_currency' or 'currency' is required for purchase events")
            return
        }

        let amount = valueToSum ?? 0.0
        
        // Validate currency code
        let validCurrencies = Locale.isoCurrencyCodes
        if !validCurrencies.contains(currency.uppercased()) {
            call.reject("Invalid currency code: \(currency)")
            return
        }

        AppEvents.shared.logPurchase(amount: amount, currency: currency.uppercased(), parameters: parameters)
        AppEvents.shared.flush()

        print("\(Self.TAG): Purchase event logged: \(eventName), amount: \(amount), currency: \(currency)")
        call.resolve()
    }

    private func logStandardEvent(eventName: String, valueToSum: Double?, parameters: [AppEvents.ParameterName: Any]) {
        let eventAppName = AppEvents.Name(eventName)
        
        if let value = valueToSum {
            AppEvents.shared.logEvent(eventAppName, valueToSum: value, parameters: parameters)
        } else {
            AppEvents.shared.logEvent(eventAppName, parameters: parameters)
        }

        AppEvents.shared.flush()
        let valueText = valueToSum != nil ? ", value: \(valueToSum!)" : ""
        print("\(Self.TAG): Event logged: \(eventName)\(valueText)")
    }

    private func ensureSDKInitialized() -> Bool {
        guard let appId = Settings.shared.appID, !appId.isEmpty else {
            print("\(Self.TAG): Facebook SDK is not initialized")
            return false
        }

        return true
    }
}

File: android/src/main/java/com/xcodeplt/facebookappevents/FacebookAppEventsPlugin.java
package com.xcodeplt.facebookappevents;

import com.android.installreferrer.BuildConfig;
import com.facebook.FacebookSdk;
import com.facebook.LoggingBehavior;
import com.facebook.appevents.AppEventsConstants;
import com.facebook.appevents.AppEventsLogger;

import android.app.Application;
import android.content.Context;
import android.os.Bundle;
import android.util.Log;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import java.util.Iterator;
import java.math.BigDecimal;
import java.util.Currency;

@CapacitorPlugin(name = "FacebookAppEvents")
public class FacebookAppEventsPlugin extends Plugin {
    private static final String TAG = "FacebookAppEvents";
    private AppEventsLogger logger;
    private boolean initialized = false;

    @PluginMethod
    public void initialize(PluginCall call) {
        String facebookAppId = call.getString("appId");
        String facebookClientToken = call.getString("clientToken");

        if (facebookAppId == null || facebookAppId.isEmpty()) {
            call.reject("Missing or empty appId");
            return;
        }

        if (facebookClientToken == null || facebookClientToken.isEmpty()) {
            call.reject("Missing or empty clientToken");
            return;
        }

        try {
            Context appContext = getContext().getApplicationContext();

            // Configure Facebook SDK
            FacebookSdk.setApplicationId(facebookAppId);
            FacebookSdk.setClientToken(facebookClientToken);
            FacebookSdk.setAutoInitEnabled(true);

            // Enable debug logging in debug builds
            if (BuildConfig.DEBUG) {
                FacebookSdk.setIsDebugEnabled(true);
                FacebookSdk.addLoggingBehavior(LoggingBehavior.APP_EVENTS);
            }

            // Initialize SDK
            FacebookSdk.sdkInitialize(appContext);
            // Create logger and activate app
            logger = AppEventsLogger.newLogger(appContext);
            AppEventsLogger.activateApp((Application) appContext);

            initialized = true;
            Log.i(TAG, "Facebook SDK initialized successfully");
            call.resolve();
        } catch (Exception e) {
            Log.e(TAG, "Failed to initialize Facebook SDK", e);
            call.reject("Failed to initialize Facebook SDK: " + e.getMessage());
        }
    }

    @PluginMethod
    public void isInitialized(PluginCall call) {
        JSObject ret = new JSObject();
        ret.put("initialized", initialized && FacebookSdk.isInitialized());
        call.resolve(ret);
    }

    @PluginMethod
    public void logEvent(PluginCall call) {
        if (!initialized) {
            call.reject("Facebook SDK not initialized. Call initialize() first.");
            return;
        }

        if (!ensureLoggerExists()) {
            call.reject("Logger not available");
            return;
        }

        String eventName = call.getString("eventName");
        Double valueToSum = call.getDouble("valueToSum");
        JSObject parameters = call.getObject("parameters");

        if (eventName == null || eventName.isEmpty()) {
            call.reject("Missing or empty 'eventName'");
            return;
        }

        try {
            Bundle bundle = createBundleFromParameters(parameters);

            // Handle standard Facebook events with special processing
            if (isPurchaseEvent(eventName)) {
                logPurchaseEvent(eventName, valueToSum, bundle, call);
            } else {
                logStandardEvent(eventName, valueToSum, bundle);
                call.resolve();
            }
        } catch (Exception e) {
            Log.e(TAG, "Error logging event: " + eventName, e);
            call.reject("Failed to log event: " + e.getMessage());
        }
    }

    private Bundle createBundleFromParameters(JSObject parameters) {
        Bundle bundle = new Bundle();

        if (parameters != null) {
            Iterator<String> keys = parameters.keys();
            while (keys.hasNext()) {
                String key = keys.next();
                Object value = parameters.opt(key);

                if (value instanceof String) {
                    bundle.putString(key, (String) value);
                } else if (value instanceof Integer) {
                    bundle.putInt(key, (Integer) value);
                } else if (value instanceof Double) {
                    bundle.putDouble(key, (Double) value);
                } else if (value instanceof Boolean) {
                    bundle.putBoolean(key, (Boolean) value);
                } else if (value instanceof Float) {
                    bundle.putFloat(key, (Float) value);
                } else if (value instanceof Long) {
                    bundle.putLong(key, (Long) value);
                } else if (value != null) {
                    // Convert other types to string as fallback
                    bundle.putString(key, value.toString());
                }
            }
        }

        return bundle;
    }

    private boolean isPurchaseEvent(String eventName) {
        return "fb_mobile_purchase".equals(eventName) || "Purchase".equals(eventName);
    }

    private void logPurchaseEvent(String eventName, Double valueToSum, Bundle bundle, PluginCall call) {
        // For purchase events, we need currency information
        String currencyCode = bundle.getString("fb_currency");
        if (currencyCode == null) {
            currencyCode = bundle.getString("currency"); // fallback
        }

        if (currencyCode == null || currencyCode.isEmpty()) {
            call.reject("'fb_currency' or 'currency' is required for purchase events");
            return;
        }

        try {
            BigDecimal amount = valueToSum != null ? BigDecimal.valueOf(valueToSum) : BigDecimal.ZERO;
            Currency currency = Currency.getInstance(currencyCode.toUpperCase());

            logger.logPurchase(amount, currency, bundle);
            logger.flush();

            Log.i(TAG, "Purchase event logged: " + eventName + ", amount: " + amount + ", currency: " + currencyCode);
            call.resolve();
        } catch (IllegalArgumentException e) {
            Log.e(TAG, "Invalid currency code: " + currencyCode, e);
            call.reject("Invalid currency code: " + currencyCode);
        }
    }

    private void logStandardEvent(String eventName, Double valueToSum, Bundle bundle) {
        if (valueToSum != null) {
            logger.logEvent(eventName, valueToSum, bundle);
        } else {
            logger.logEvent(eventName, bundle);
        }

        logger.flush();
        Log.i(TAG, "Event logged: " + eventName + (valueToSum != null ? ", value: " + valueToSum : ""));
    }

    private boolean ensureLoggerExists() {
        if (!FacebookSdk.isInitialized()) {
            Log.w(TAG, "Facebook SDK is not initialized");
            return false;
        }

        if (logger == null) {
            try {
                logger = AppEventsLogger.newLogger(getContext());
                Log.i(TAG, "AppEventsLogger instance created");
            } catch (Exception e) {
                Log.e(TAG, "Failed to create AppEventsLogger", e);
                return false;
            }
        }

        return true;
    }
}
